src/common/TimeZoneUtil.cpp

```diff
@@ -70,9 +83,26 @@ namespace
 			return unicodeName.begin();
 		}
 
+		UCalendar* getCalendar(const Jrd::UnicodeUtil::ConversionICU& icuLib, UErrorCode* err) const
+		{
+			UCalendar* val = calendar.load();
+			if (!val)
+			{
+				val = icuLib.ucalOpen(getUnicodeName(), -1, NULL, UCAL_GREGORIAN, err);
+				if (!val)
+					return val;
+
+				UCalendar* old = calendar.exchange(val);
+				if (old)
+					return old;
+			}
+			return icuLib.ucalClone(val, err);
```

Comment:
`ucal_clone` is not documented to be thread-safe, so it should not be assumed to be.
I do also not know its performance.
I'd instead implement a pool of unused objects where concurrent usage may do open/close additional copies and single usage would be reverted to the pool after its usage.

Reply:
I followed this docs

https://unicode-org.github.io/icu/userguide/icu/design.html#thread-safe-const-apis

https://unicode-org.github.io/icu/userguide/icu/design.html#clone-vs-open

Reply:
About performance, with patch I have following result:

| TEST | DURATION |
|--------|--------|
TZ -> TS	| 44
TS -> TZ	| 70
TS -> TS	| 13
TZ -> TZ	| 13

Reply:
"However, you cannot use a reference to an open service object in two threads at the same time if either of them calls any non-const API."

`ucal_clone` is const, but `ucal_setmillis` is not.

In no place it says `ucal_clone` can be called concurrently with others non-const functions, so it's not safe.

Reply:
`ucal_clone` is const by itself thus it could be called with no sync.
`ucal_setmillis` is non-const but it uses own private instance of UCalendar, cloned from existing one.

Reply:
Also, the full cite is:

> However, you cannot use a reference to an open service object in two threads at the same time if either of them calls any non-const API. An individual open service object is not thread-safe for concurrent “writes”. Rather, for non-const use, you must use the clone function to create a copy of the service you want and then pass this copy to the second thread. This procedure allows you to use the same service in different threads, but avoids any thread synchronization or deadlock problems.